### PR TITLE
Make PortSockTest::poll_functionality_many_sockets more robust

### PR DESCRIPTION
**Pull request by dnakamura**
Adjust logic to make less assumptions about socket implementation,
and more tolerant of unexpected conditions.

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

### CODE REVIEW

**Comment by dnakamura** [comment]
@genie-omr  build all

**Comment by dnakamura** [comment]
@genie-omr build all

**Comment by dnakamura** [comment]
See also #5581

**Comment by 0xdaryl** [comment]
@rwy0717 or @youngar : would one of you be able to review this please?

**Comment by 0xdaryl** [comment]
Also tagging @tajila and @babsingh for possible reviews.  This is somewhat of a higher priority for OMR as that macOS port lib test fails repeatedly as part of CI validation of a PR.

**Comment by tajila** [comment]
@babsingh Can you review this since you have reviewed socket code in the past

File: fvtest/porttest/omrsockTest.cpp [hunk 1]
```diff
@@ -1285,26 +1288,33 @@ TEST(PortSockTest, poll_functionality_many_sockets)
...
-	for (int32_t i = 0; i < 10; i++) {
-		OMRPORTLIB->sock_get_pollfd_info(OMRPORTLIB, &pollArray[i], &socketPoll, &revents);
-		if (0 != (revents & OMRSOCK_POLLOUT)) {
-			ASSERT_EQ(clientSocket, socketPoll);
-			bytesSent = OMRPORTLIB->sock_send(OMRPORTLIB, clientSocket, (uint8_t *)msg, strlen(msg) + 1, 0);
-			ASSERT_GT(bytesSent, 0);
-			break;
-		}
-	}
+	/* The client socket should be ready to write */
```
Comment (babsingh) [inline fvtest/porttest/omrsockTest.cpp:1291]:
```suggestion
	/* The client socket should be ready to write. */
```

File: fvtest/porttest/omrsockTest.cpp [hunk 2]
```diff
@@ -1285,26 +1288,33 @@ TEST(PortSockTest, poll_functionality_many_sockets)
...
+	OMRPORTLIB->sock_get_pollfd_info(OMRPORTLIB, &pollArray[CLIENT_SOCKET_POLL_IDX], &socketPoll, &revents);
+	ASSERT_EQ(clientSocket, socketPoll);
+	ASSERT_NE(revents & OMRSOCK_POLLOUT, 0);
+
+	bytesSent = OMRPORTLIB->sock_send(OMRPORTLIB, clientSocket, (uint8_t *)msg, strlen(msg) + 1, 0);
+	ASSERT_GT(bytesSent, 0);
+
+
+	/* Ignore OMRSOCK_POLLOUT on client socket, since its state is indeterminate/ */
```
Comment (babsingh) [inline fvtest/porttest/omrsockTest.cpp:1300]: Multi-line comment:
```suggestion
	/* Ignore OMRSOCK_POLLOUT on client socket, since its state is indeterminate.
```

File: fvtest/porttest/omrsockTest.cpp [hunk 3]
```diff
@@ -1285,26 +1288,33 @@ TEST(PortSockTest, poll_functionality_many_sockets)
...
+	/* Listen for HUP, since an error is returned if sock_pollfd_init recieves an event arg of 0. */
```
Comment (babsingh) [inline fvtest/porttest/omrsockTest.cpp:1301]: Multi-line comment:
```suggestion
	 * Listen for HUP, since an error is returned if sock_pollfd_init recieves an event arg of 0.
	 */
```

File: fvtest/porttest/omrsockTest.cpp [hunk 4]
```diff
@@ -1285,26 +1288,33 @@ TEST(PortSockTest, poll_functionality_many_sockets)
...
 	/* Check that server POLLIN is ready. Give poll up to 10 times to be ready. */
+	bool succeeded = false;
 	for (int32_t i = 0; i < 10; i++) {
-		if (2 == (rc = OMRPORTLIB->sock_poll(OMRPORTLIB, pollArray, 10, 1000))) {
-			OMRPORTLIB->sock_get_pollfd_info(OMRPORTLIB, &pollArray[8], &socketPoll, &revents);
-			ASSERT_EQ(connectedServerSocket, socketPoll);
-			ASSERT_NE((revents & OMRSOCK_POLLIN), 0);
-			break;
+		rc = OMRPORTLIB->sock_poll(OMRPORTLIB, pollArray, 10, 1000);
+		if(rc > 0){
```
Comment (babsingh) [inline fvtest/porttest/omrsockTest.cpp:1308]:
```suggestion
		if (rc > 0) {
```

File: fvtest/porttest/omrsockTest.cpp [hunk 5]
```diff
@@ -1285,26 +1288,33 @@ TEST(PortSockTest, poll_functionality_many_sockets)
...
+			OMRPORTLIB->sock_get_pollfd_info(OMRPORTLIB, &pollArray[SERVER_SOCKET_POLL_IDX], &socketPoll, &revents);
+			ASSERT_EQ(socketPoll, connectedServerSocket);
+			if(0 != (revents & OMRSOCK_POLLIN)){
```
Comment (babsingh) [inline fvtest/porttest/omrsockTest.cpp:1311]:
```suggestion
			if (0 != (revents & OMRSOCK_POLLIN)) {
```

**Review by babsingh — CHANGES_REQUESTED** [review]
The code looks functionally good. Only nitpicked minor formatting issues.

The build failures happen due to failure of `git` operations (infra).

The commit message should also address this change as a fix for:
`macOS port lib test fails repeatedly as part of CI validation of a PR`.

File: fvtest/porttest/omrsockTest.cpp [hunk 6]
```diff
@@ -1285,26 +1288,33 @@ TEST(PortSockTest, poll_functionality_many_sockets)
...
+
+
+	/* Ignore OMRSOCK_POLLOUT on client socket, since its state is indeterminate/ */
+	/* Listen for HUP, since an error is returned if sock_pollfd_init recieves an event arg of 0. */
+	 EXPECT_EQ(OMRPORTLIB->sock_pollfd_init(OMRPORTLIB, &pollArray[CLIENT_SOCKET_POLL_IDX], clientSocket, OMRSOCK_POLLHUP), 0);
 
 	/* Check that server POLLIN is ready. Give poll up to 10 times to be ready. */
+	bool succeeded = false;
 	for (int32_t i = 0; i < 10; i++) {
-		if (2 == (rc = OMRPORTLIB->sock_poll(OMRPORTLIB, pollArray, 10, 1000))) {
```
Comment (babsingh) [inline fvtest/porttest/omrsockTest.cpp:1300]: `sock_poll` returns the number of sockets ready for IO. in the new changes, `(rc == 2)` is changed to `(rc > 0)`. why was this change required? should the test fail for `(rc == 1)`?

Comment (dnakamura) [reply fvtest/porttest/omrsockTest.cpp:1300]: The client socket isn't guaranteed to be ready to write again. In addition, its possible that the OS is smart enough to skip the network stack and the socket buffers are shared (eg client write buffer == server read buffer). If that is the case, and the buffer is full, the client socket will not be ready to write again until the server reads to free up space in the buffer.

**Comment by dnakamura** [comment]
@genie-omr build all

**Review by babsingh — APPROVED** [review]

**Review by rwy0717 — APPROVED** [review]
